alibi_detect/utils/fetching/fetching.py

```diff
@@ -1,40 +1,57 @@
 import logging
 import os
-from urllib.parse import urljoin, quote_plus
+from io import BytesIO
 from pathlib import Path
+from typing import Union, TYPE_CHECKING, Tuple
 import dill
-import tensorflow as tf
-from tensorflow.python.keras import backend
-from typing import Tuple, Union, List
-from io import BytesIO
+
 import requests
 from requests import RequestException
-from alibi_detect.base import BaseDetector
-from alibi_detect.ad import AdversarialAE, ModelDistillation
+
+import tensorflow as tf
+from tensorflow.python.keras import backend
+
 from alibi_detect.models.tensorflow import PixelCNN
-from alibi_detect.od import (IForest, LLR, Mahalanobis, OutlierAE, OutlierAEGMM, OutlierProphet,
-                             OutlierSeq2Seq, OutlierVAE, OutlierVAEGMM, SpectralResidual)
-from alibi_detect.utils.saving import load_detector  # type: ignore
+
+if TYPE_CHECKING:
+    # Import the true objects directly for typechecking. If we import at the module level the import_optional method
+    # replaces them with all type.
+    from alibi_detect.ad.adversarialae import AdversarialAE  # noqa
+    from alibi_detect.ad.model_distillation import ModelDistillation  # noqa
+    from alibi_detect.base import BaseDetector  # noqa
+    from alibi_detect.od.llr import LLR  # noqa
+    from alibi_detect.od.isolationforest import IForest  # noqa
+    from alibi_detect.od.mahalanobis import Mahalanobis  # noqa
+    from alibi_detect.od.aegmm import OutlierAEGMM  # noqa
+    from alibi_detect.od.ae import OutlierAE  # noqa
+    from alibi_detect.od.prophet import OutlierProphet  # noqa
+    from alibi_detect.od.seq2seq import OutlierSeq2Seq  # noqa
+    from alibi_detect.od.vae import OutlierVAE  # noqa
+    from alibi_detect.od.vaegmm import OutlierVAEGMM  # noqa
+    from alibi_detect.od.sr import SpectralResidual  # noqa
+
+from alibi_detect.utils.fetching.url import _join_url
+from alibi_detect.utils.saving import load_detector
 
 # do not extend pickle dispatch table so as not to change pickle behaviour
 dill.extend(use_dill=False)
 
 logger = logging.getLogger(__name__)
 
 Data = Union[
-    BaseDetector,
-    AdversarialAE,
-    ModelDistillation,
-    IForest,
-    LLR,
-    Mahalanobis,
-    OutlierAEGMM,
-    OutlierAE,
-    OutlierProphet,
-    OutlierSeq2Seq,
-    OutlierVAE,
-    OutlierVAEGMM,
-    SpectralResidual
+    'BaseDetector',
```

Comment:
What happens if forward refs aren't used?

Reply:
The logic is a bit complex. If we remove the forward refs:
1. If not `TYPE_CHECKING` these objects won't be imported so an error will be thrown as they're not defined.
2. If we remove the import conditional on `TYPE_CHECKING` then if the user doesn't have all the optional dependencies installed anything imported that depends on an optional dependency will fail and throw an error.
3. If we use optional import instead (i.e. import from the `__init__.py`) then If the user doesn't have all optional dependencies installed then import optional will replace a `Class` Object with an instance of the `MissingDependency` class. This will throw an error when passed to `Union`. 
4. As well as this the import optional return type has to be `Any` and so this basically makes any type checking meaningless.

Reply:
Ah yes I remember now, thanks!